lib/lightning/attempts/handlers.ex

```diff
@@ -169,10 +169,18 @@ defmodule Lightning.Attempts.Handlers do
 
     defp update(complete_run) do
       Repo.transact(fn ->
-        with run = %Run{} <-
-               get_run(complete_run.run_id) || {:error, :not_found},
-             {:ok, _} <- to_dataclip(complete_run) |> Repo.insert() do
-          update_run(run, complete_run)
+        case get_run(complete_run.run_id) do
+          nil ->
+            # Return an error changeset when the run is not found
+            %Run{}
+            |> change()
+            |> put_change(:exit_reason, "error")
+            |> put_change(:error_type, "Run not found")
```

Comment:
I'm not sure this is the way to go about it. In the case of an error we don't return a Run changeset, we return a `CompleteRun` changeset, it's purpose is to stand in between the worker sending a bunch of data that represents data that is for a run and a dataclip. It can be thought of as a 'command'.

My suggestion is to add a `run: Run` field to the schema, we don't cast it - but set the field and then validate it.

```elixir
  # ... ^ new/1
  |> then(&assign_run/1)
end

defp assign_run(changeset) do
  with %{valid?: true} <- changeset,
        %Run{} = run <-
          get_field(changeset, :run_id) |> get_run() || :not_found do
    put_change(changeset, :run, run)
  else
    :not_found ->
      add_error(changeset, :run_id, "not found")

    changeset ->
      changeset
  end
end
```

And then further down the `update/1` function can assume the changeset is valid, and get the `run` off the changeset instead of finding it itself.

Reply:
The proposal is to be very granular with the fix changing only the flow of the complete run.
The change consists of just adding the changeset to the return value.